### PR TITLE
Module_5 Submission

### PR DESCRIPTION
**Pull request by pjrice4**
I spent several hours trying to get Mongoose to install and work, but
did not have any luck.  I uninstalled all the relevant parts and
reinstalled.  Hopefully my files will run.  I might try to track down a
Mac this afternoon to see if that will do the trick.

Thanks,

Patrick

### CODE REVIEW

**Comment by jamdon** [comment]
Hi Patrick,

What errors are you getting when you try to install mongoose on Windows?

James

**Comment by pjrice4** [comment]
James,

Here is the output I'm getting when I try to npm install.  Unfortunately
this has gone slightly beyond my computer knowledge, and there are so many
possible solutions on Stackoverflow that I can't keep it straight.

---

Your environment has been set up for using Node.js 0.12.4 (ia32) and npm.

C:\Users\Patrick>cd documents

C:\Users\Patrick\Documents>cd github

C:\Users\Patrick\Documents\GitHub>cd module_5

C:\Users\Patrick\Documents\GitHub\Module_5>npm install mongoose --save
/

> kerberos@0.0.12 install
> C:\Users\Patrick\Documents\GitHub\Module_5\node_module
> s\mongoose\node_modules\mongodb\node_modules\mongodb-core\node_modules\kerberos
> (node-gyp rebuild 2> builderror.log) || (exit 0)

C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose\node_modules\mo
ngodb\node_modules\mongodb-core\node_modules\kerberos>if not defined
npm_config_
node_gyp (node "C:\Program Files
(x86)\nodejs\node_modules\npm\bin\node-gyp-bin\
....\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (rebuild)
Building the projects in this solution one at a time. To enable parallel
build,
please add the "/m" switch.
\  kerberos.cc
  security_buffer.cc
  security_buffer_descriptor.cc
C:\Program Files (x86)\Microsoft Visual Studio
10.0\VC\include\xlocale(323): wa
rning C4530: C++ exception handler used, but unwind semantics are not
enabled.
Specify /EHsc
[C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose
\node_modules\mongodb\node_modules\mongodb-core\node_modules\kerberos\build\ker
beros.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio
10.0\VC\include\xlocale(323): wa
rning C4530: C++ exception handler used, but unwind semantics are not
enabled.
Specify /EHsc
[C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose
\node_modules\mongodb\node_modules\mongodb-core\node_modules\kerberos\build\ker
beros.vcxproj]
  security_context.cc
  security_credentials.cc
C:\Program Files (x86)\Microsoft Visual Studio
10.0\VC\include\xlocale(323): wa
rning C4530: C++ exception handler used, but unwind semantics are not
enabled.
Specify /EHsc
[C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose
\node_modules\mongodb\node_modules\mongodb-core\node_modules\kerberos\build\ker
beros.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio
10.0\VC\include\xlocale(323): wa
rning C4530: C++ exception handler used, but unwind semantics are not
enabled.
Specify /EHsc
[C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose
\node_modules\mongodb\node_modules\mongodb-core\node_modules\kerberos\build\ker
beros.vcxproj]
  base64.c
  kerberos_sspi.c
     Creating library
C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\m

ongoose\node_modules\mongodb\node_modules\mongodb-core\node_modules\kerberos\
  build\Release\kerberos.lib and object
C:\Users\Patrick\Documents\GitHub\Modul

e_5\node_modules\mongoose\node_modules\mongodb\node_modules\mongodb-core\node
  _modules\kerberos\build\Release\kerberos.exp
  Generating code
  Finished generating code
  kerberos.vcxproj ->
C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\m

ongoose\node_modules\mongodb\node_modules\mongodb-core\node_modules\kerberos\
  build\Release\kerberos.node
npm WARN installMany nopt was bundled with node-pre-gyp@0.6.4, but bundled
packa
ge wasn't found in unpacked tree
npm WARN installMany npmlog was bundled with node-pre-gyp@0.6.4, but
bundled pac
kage wasn't found in unpacked tree
npm WARN installMany request was bundled with node-pre-gyp@0.6.4, but
bundled pa
ckage wasn't found in unpacked tree
npm WARN installMany semver was bundled with node-pre-gyp@0.6.4, but
bundled pac
kage wasn't found in unpacked tree
npm WARN installMany tar was bundled with node-pre-gyp@0.6.4, but bundled
packag
e wasn't found in unpacked tree
npm WARN installMany tar-pack was bundled with node-pre-gyp@0.6.4, but
bundled p
ackage wasn't found in unpacked tree
npm WARN installMany mkdirp was bundled with node-pre-gyp@0.6.4, but
bundled pac
kage wasn't found in unpacked tree
npm WARN installMany rc was bundled with node-pre-gyp@0.6.4, but bundled
package
 wasn't found in unpacked tree
npm WARN installMany rimraf was bundled with node-pre-gyp@0.6.4, but
bundled pac
kage wasn't found in unpacked tree
/

> bson-ext@0.1.7 install
> C:\Users\Patrick\Documents\GitHub\Module_5\node_modules
> \mongoose\node_modules\bson\node_modules\bson-ext
> (node-pre-gyp install --fallback-to-build) || (node-gyp rebuild 2>
> builderror.
> log) || (exit 0)

C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose\node_modules\bs
on\node_modules\bson-ext>if not defined npm_config_node_gyp (node
"C:\Program Fi
les
(x86)\nodejs\node_modules\npm\bin\node-gyp-bin\....\node_modules\node-gyp\
bin\node-gyp.js" clean )  else (clean)

C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose\node_modules\bs
on\node_modules\bson-ext>if not defined npm_config_node_gyp (node
"C:\Program Fi
les
(x86)\nodejs\node_modules\npm\bin\node-gyp-bin\....\node_modules\node-gyp\
bin\node-gyp.js" configure --fallback-to-build
--module=C:\Users\Patrick\Documen
ts\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_modules\bson-ext
\build\Release\bson.node --module_name=bson
--module_path=C:\Users\Patrick\Docum
ents\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_modules\bson-e
xt\build\Release )  else (configure --fallback-to-build
--module=C:\Users\Patric
k\Documents\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_modules
\bson-ext\build\Release\bson.node --module_name=bson
--module_path=C:\Users\Patr
ick\Documents\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_modul
es\bson-ext\build\Release )

C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose\node_modules\bs
on\node_modules\bson-ext>if not defined npm_config_node_gyp (node
"C:\Program Fi
les
(x86)\nodejs\node_modules\npm\bin\node-gyp-bin\....\node_modules\node-gyp\
bin\node-gyp.js" build --fallback-to-build
--module=C:\Users\Patrick\Documents\G
itHub\Module_5\node_modules\mongoose\node_modules\bson\node_modules\bson-ext\bui
ld\Release\bson.node --module_name=bson
--module_path=C:\Users\Patrick\Documents
\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_modules\bson-ext\b
uild\Release )  else (build --fallback-to-build
--module=C:\Users\Patrick\Docume
nts\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_modules\bson-ex
t\build\Release\bson.node --module_name=bson
--module_path=C:\Users\Patrick\Docu
ments\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_modules\bson-
ext\build\Release )
Building the projects in this solution one at a time. To enable parallel
build,
please add the "/m" switch.
  bson.cc
c:\users\patrick\documents\github\module_5\node_modules\mongoose\node_modules\b
son\node_modules\bson-ext\node_modules\nan\nan_new.h(208): warning C4244:
'argu
ment' : conversion from 'int64_t' to 'double', possible loss of data
[C:\Users\
Patrick\Documents\GitHub\Module_5\node_modules\mongoose\node_modules\bson\node_
modules\bson-ext\build\bson.vcxproj]
          ..\ext\bson.cc(589) : see reference to function template
instantiatio
  n 'v8::Local<T> NanNewv8::Number,int64_t(A0)' being compiled
          with
          [
              T=v8::Number,
              A0=int64_t
          ]
     Creating library
C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\m
  ongoose\node_modules\bson\node_modules\bson-ext\build\Release\bson.lib
and ob
  ject
C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongoose\node_mo
  dules\bson\node_modules\bson-ext\build\Release\bson.exp
  Generating code
  Finished generating code
  bson.vcxproj ->
C:\Users\Patrick\Documents\GitHub\Module_5\node_modules\mongo
  ose\node_modules\bson\node_modules\bson-ext\build\Release\bson.node
mongoose@4.0.5 node_modules\mongoose
├── regexp-clone@0.0.1
├── sliced@0.0.5
├── muri@1.0.0
├── hooks-fixed@1.0.1
├── mpromise@0.5.4
├── kareem@1.0.1
├── mpath@0.1.1
├── async@0.9.0
├── ms@0.1.0
├── mquery@1.5.1 (debug@0.7.4, bluebird@2.9.9)
├── mongodb@2.0.33 (readable-stream@1.0.31, mongodb-core@1.1.32)
└── bson@0.3.2 (bson-ext@0.1.7)

C:\Users\Patrick\Documents\GitHub\Module_5>

Patrick J. Rice
Email: PJRice4@gmail.com
Cell: 425.753.2610

**Comment by jamdon** [comment]
It's hard to tell which part is the error in that spew. Is some of the
output in red?

James

**Comment by pjrice4** [comment]
Thank you for your help with this, I thought I had figured it out in class
the other day, but it appears not.
[image: Inline image 2][image: Inline image 3]

Thanks,

Patrick

Patrick J. Rice
Email: PJRice4@gmail.com
Cell: 425.753.2610

**Comment by pjrice4** [comment]
James,

I've gone through all the steps again and still can't get it to straighten
out.  I am able to get the mongoDB up and running, and I can do the simple
queries to the trucksJSON file, so I know that is working.  I can get the
node server up and running, I just can't get the two to talk to each other.

I submitted my assignment, so there is at least something to grade,
hopefully there is enough there to get by.  I'm going to try to borrow a
friend's Mac this afternoon, to see if I can get it up and running, but I
don't know if that will come through.

Either way, thank you for all your help this quarter.

Thanks,

Patrick

Patrick J. Rice
Email: PJRice4@gmail.com
Cell: 425.753.2610

File: routes/truckRoutes.js [hunk 1]
```diff
@@ -1,3 +1,60 @@
...
+router.route('/')
+	.get(function(request, response) {
+		Truck.find(function(error, trucks) {
+			if(error) {
+				response.status(500).send(error);
+			} else {
+				response.send(trucks);
+			}
+		});
+	});
```
Comment (jamdon) [inline routes/truckRoutes.js:19]: When you chain route handlers, you don't want to terminate the ones preceding others with a semi-colon.

File: routes/truckRoutes.js [hunk 2]
```diff
@@ -1,3 +1,60 @@
...
+	.get(function(request, response) {
+		var id = request.params.id;
+		Truck.findById(id, function(error, truck) {
+			if (error) {
+				response.status(500).send(error);
+			} else {
+				response.send(truck);
+			}
+		});
+	});
```
Comment (jamdon) [inline routes/truckRoutes.js:42]: Same thing here as above - don't terminate the chain if there are other statements chaining off of this one.

File: models/truckModel.js [hunk 3]
```diff
@@ -2,14 +2,19 @@
 // when exporting the mongoose model you will create using this schema, 
 // name your model 'Truck' (see the example in the slide deck for 'Book')
 // this is to match the name of the collection the data was imported into
+var mongoose = require('mongoose');
+var Schema = mogoose.Schema;
```
Comment (jamdon) [inline models/truckModel.js:6]: spelling error - should be mongoose

File: routes/truckRoutes.js [hunk 4]
```diff
@@ -1,3 +1,60 @@
...
+						response.status(500).send(error);
+					} else {
+						response.sendStatus(200);
+					}
+				});
+			}
+		});
+	});
+
+module.exports = truckRouter;
```
Comment (jamdon) [inline routes/truckRoutes.js:60]: Your variable above is router, but you're trying to export truckRouter here. Instead `module.exports = router;`

File: routes/truckRoutes.js [hunk 5]
```diff
@@ -1,3 +1,60 @@
...
+				response.status(500).send(error);
+			} else {
+				response.send(trucks);
+			}
+		});
+	});
+	.post(function(request, response) {
+		var newTruck = request.body;
+		var truck = new Truck(newTruck);
+		newTruck.save(function(error, truck) {
```
Comment (jamdon) [inline routes/truckRoutes.js:23]: newTruck is the request.body at this point, and truck is the Truck object. So instead, you'll want to call `truck.save()` here

**Comment by jamdon** [comment]
Worked once I addressed the issues I commented on inline. :+1:

**Pull request closed without merging**